### PR TITLE
Mapped property converters + Data only converters

### PR DESCRIPTION
**Pull request by dstepanov**
Solving the problem with conflicting converters by introducing `DataConversionService`. Fixing #973

Adding a way to convert one mapper property to another just like JPA's `AttributeConverter`.
Proper fix for https://github.com/micronaut-projects/micronaut-data/issues/832

### CODE REVIEW

File: data-runtime/src/main/java/io/micronaut/data/runtime/support/convert/DefaultTypeConverterRegistry.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+@Internal
+final class DefaultTypeConverterRegistry implements TypeConverterRegistry {
+
+    private final BeanLocator beanLocator;
+    private final List<TypeConverterProvider> typeConverterTransformers;
+
+    DefaultTypeConverterRegistry(BeanLocator beanLocator) {
+        this.beanLocator = beanLocator;
+        List<TypeConverterProvider> typeConverterTransformers = new ArrayList<>();
+        SoftServiceLoader.load(TypeConverterProvider.class).collectAll(typeConverterTransformers);
```
Comment (graemerocher) [inline data-runtime/src/main/java/io/micronaut/data/runtime/support/convert/DefaultTypeConverterRegistry.java:41]: Do we need to use service loader here? Could `TypeConverterProvider` instances not be beans with `@Requires(classes=..)`?

**Review by graemerocher — CHANGES_REQUESTED** [review]

File: data-runtime/src/main/java/io/micronaut/data/runtime/support/convert/DefaultTypeConverterRegistry.java (hunk 1 above)
Comment (dstepanov) [reply data-runtime/src/main/java/io/micronaut/data/runtime/support/convert/DefaultTypeConverterRegistry.java]: That was probably overkill

**Comment by dstepanov** [comment]
Not sure if `TypeConverter` is a good name, we already have `TypeConverter` in core

**Comment by graemerocher** [comment]
Could just call it `AttributeConverter`

File: data-model/src/main/java/io/micronaut/data/model/runtime/convert/AttributeConverter.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * The attribute converter is used for converting mapped entity value to the persisted value and back.
+ *
+ * @param <X> The entity value type
+ * @param <Y> The persisted value type
+ * @author Denis Stepanov
+ * @since 3.1
+ */
+public interface AttributeConverter<X, Y> {
```
Comment (graemerocher) [inline data-model/src/main/java/io/micronaut/data/model/runtime/convert/AttributeConverter.java:30]: Maybe add `@Indexed(AttributeConverter.class)` to speed lookup

**Review by graemerocher — APPROVED** [review]

**Pull request merged**